### PR TITLE
Added custom ethereum network url generation

### PR DESCRIPTION
**Pull request by BboyAkers**
This should allow for the ability to generate urls for custom networks based upon [EIP 3091](https://github.com/ethereum/EIPs/blob/master/EIPS/eip-3091.md).

Still a WIP. Tried several other ways to do this but decided on this. Would love to know ya'lls thoughts 🙂!! Any better ideas to on building this out? @Gudahtt  @rekmarks @danfinlay @darkwing

### CODE REVIEW

File: src/account-link.ts
Comment (Gudahtt) [inline src/account-link.ts:3]: It is a bit unfortunate that `network` is still a required parameter here even when the caller is resolving a custom network.

Maybe resolving a custom network should be a separate function altogether? It doesn't share any common code with resolving a built-in network.

File: src/account-link.ts
Comment (Gudahtt) [inline src/account-link.ts:5]: It would be better to use the `URL` API to parse the URL, so that you can ensure this path still makes sense. This would be problem if the custom network URL passed in included query string parameters or a hash, or if it ended in a slash already.

With the URL API you can deconstruct this into its pieces, check the pathname for a trailing slash and add one only if it's missing, add the rest of the path, then reconstruct the URL again preserving all query string parameters. That would allow the caller to pass in important information like the API key, for any block explorer that accepts it via query string anyway.

File: src/token-tracker-link.ts
Comment (Gudahtt) [inline src/token-tracker-link.ts:11]: The last part of this (`?a=${ holderAddress }`) isn't part of the EIP. Maybe it's best that we left this out for now.

For custom networks we can just use the account link instead, and rely upon the user to find their token transactions from there.

**Review by Gudahtt — CHANGES_REQUESTED** [review]
Thanks, this is a good start!

I think the API would be simpler if we separated the functions based upon EIP-3091 from the others. Maybe they could all go in a separate module or directory or something/

Also it would be great to have a couple of usage examples in the README

File: src/account-link.ts
Comment (BboyAkers) [reply src/account-link.ts:3]: So would we call that separate function in a conditional statement or are we exporting that separate function all together?

Comment (BboyAkers) [reply src/account-link.ts:3]: My concern is who will dynamically determine what is a custom url or not. I feel as though we should just take a string and based upon what we have defined we determine what is and isn't a custom url. I could be misinterpreting what you're stating.

Comment (Gudahtt) [reply src/account-link.ts:3]: The caller of this function already has to know whether they're on a custom network or not. They have to know what `network` means, and what `customNetworkUrl` means, and when to pass in each.

I just thought this would be simpler to use of the two cases were separated, such that `getAccountLink(address: string, network: string)` would stay as-is, and we'd have a new function like `customBlockExplorer.getAccountLink(address: string, customNetworkUrl: string)` for the custom network case. Then it's clear for the user of the library what each function does, and there are no extraneous parameters, and the body of the function doesn't end up being two completely distinct branches.

**Comment by BboyAkers** [comment]
Made some more changes! Let me know if these work 🙂 . If not, please do yo thang 😎 **ThePRMaster**

File: src/token-tracker-link.ts
Comment (Gudahtt) [inline src/token-tracker-link.ts:13]: There's something off about the indentation here :thinking:

**Comment by brad-decker** [comment]
@BboyAkers - could I get you to review #32 and see how you think our work should best merge when it comes time?

File: src/token-tracker-link.ts
Comment (Gudahtt) [inline src/token-tracker-link.ts:10]: Unfortunately it's not quite this easy. The "base" URL passed to the URL constructor as the second param is only used for its origin. Any path or query parameters are discarded.

I'd suggest something like this:

```suggestion
    const { username, password, protocol, host, pathname, search, hash } = new URL(customNetworkUrl)
    const newPath = pathname.endsWith('/') ? `${pathname}token/${tokenAddress}` : `${pathname}/token/${tokenAddress}` 
    const auth = username ? `${username}:${password}` : ''
    let parsedUrl = new URL(`${protocol}//${auth}${host}${newPath}${search}${hash}`)
```

That should preserve all pieces of the URL while adding a path.

Check out this diagram for a reference of all of these different URL parts: https://nodejs.org/api/url.html#url_url_strings_and_url_objects
I found it helpful.

Comment (Gudahtt) [reply src/token-tracker-link.ts:10]: Oh and the same advice applies to getCustomBlockExplorerAccountLink and getExplorerLink as well.

Maybe this URL manipulation can be moved to a common function :thinking: Something like `addPathToUrl`

File: src/index.ts
Comment (Gudahtt) [inline src/index.ts:3]: Nit: We could rename the functions `createAccountLink` and `createCustomAccountLink` respectively, to avoid having to alias them here

**Comment by Gudahtt** [comment]
If the custom network functionality is isolated to separate functions (like `getCustomBlockExplorerAccountLink`), I don't think there will be any overlap. Your PR will ensure the etherscan links are chain ID based, and custom networks won't use a chain ID or network ID.

**Comment by BboyAkers** [comment]
> @BboyAkers - could I get you to review #32 and see how you think our work should best merge when it comes time? 

I'll take a look today 🙂

**Comment by brad-decker** [comment]
> > @BboyAkers - could I get you to review #32 and see how you think our work should best merge when it comes time? 
> 
> I'll take a look today 🙂

Still want your review, merged but hasn't been released yet. If you get time to look at it I'll make any suggested changes ahead of the package update. Maybe holding off until this lands too :)

**Comment by BboyAkers** [comment]
> > > @BboyAkers - could I get you to review #32 and see how you think our work should best merge when it comes time? 
> > 
> > I'll take a look today 🙂
> 
> Still want your review, merged but hasn't been released yet. If you get time to look at it I'll make any suggested changes ahead of the package update. Maybe holding off until this lands too :) 

Gotcha! Just logged off for the day 😃. I should be able to give it in the next hour or so!

**Comment by BboyAkers** [comment]
Pulling locally to test some things and fix some wonkyness on my merge conflict. Should hopefully be by morning. 😅

File: src/token-tracker-link.ts
Comment (BboyAkers) [reply src/token-tracker-link.ts:11]: Thank you for showing me this 🙂 was trying to work with the URL API and I had a feeling I wasn't using it the way I really wanted to lol. The docs on it weren't the most amazing imo 😅

Comment (Gudahtt) [reply src/token-tracker-link.ts:11]: I had the same experience! The MDN docs are _OK_ but it's really hard to tell what it does overall without a nice diagram like this. This one I linked was meant to be a comparison between the Node.js URL API and the standard WHATWG URL API, rather than a reference for either. But even still I always refer back to it because it's so much easier to understand than the docs.

**Comment by BboyAkers** [comment]
@Gudahtt @brad-decker  let me know what you think! 🙂

File: src/token-tracker-link.ts
Comment (brad-decker) [reply src/token-tracker-link.ts:11]: After reviewing I don't believe this comment was addressed unless I'm missing something. I think we should add some test cases for URLs that contain paths, auth or query parameters. Those tests can then be grouped to test the recommended utility function `addPathToUrl`

Comment (BboyAkers) [reply src/token-tracker-link.ts:11]: Whoops! Thanks for the heads up and gotcha!

File: src/account-link.ts
Comment (Gudahtt) [inline src/account-link.ts:18]: Nit: The indentation is off here (we really need a linter in this repo :sweat_smile: ). Also the trailing newline was removed. I've swapped out `let` for `const` as well, since the value is never re-assigned.

```suggestion
  const parsedUrl = addPathToUrl(customNetworkUrl, 'address', address)
  return parsedUrl
}

```

File: src/helpers.ts
Comment (Gudahtt) [inline src/helpers.ts:11]: Nit: Indentation, newline, and use of `let` for a `const` here as well. I've fixed these, and removed trailing whitespace on the empty lines:
```suggestion
  const { username, password, protocol, host, pathname, search, hash } = new URL(customNetworkUrl)

  const newPath = pathname.endsWith('/') ? `${pathname}${linkType}/${suffixType}` : `${pathname}/${linkType}/${suffixType}` 

  const auth = username ? `${username}:${password}` : ''

  const parsedUrl = new URL(`${protocol}//${auth}${host}${newPath}${search}${hash}`)

  return parsedUrl.toString()
}

```

File: src/token-tracker-link.ts
Comment (Gudahtt) [inline src/token-tracker-link.ts:12]: Nit:
```suggestion
    const parsedUrl = addPathToUrl(customNetworkUrl, 'token', tokenAddress)
```

File: src/explorer-link.ts
Comment (Gudahtt) [inline src/explorer-link.ts:8]: Nit: Inconsistent indentation, and use of `let` over `const`:
```suggestion
  const parsedUrl = addPathToUrl(customNetworkUrl, 'tx', hash) 
  return parsedUrl
}
```

**Review by Gudahtt — COMMENTED** [review]
Lots of indentation issues! I added a bunch of suggestions - I'll commit them all in a batch to move the review forward.

File: src/token-tracker-link.ts
Comment (Gudahtt) [inline src/token-tracker-link.ts:20]: Oh, another indentation issue here:
```suggestion
  if (customNetworkUrl?.length){
    const parsedUrl = addPathToUrl(customNetworkUrl, 'token', tokenAddress)
    return parsedUrl
  } else {
    const prefix = prefixForNetwork(network)
    return prefix !== null ? 
      `https://${prefix}etherscan.io/token/${tokenAddress}${ 
        holderAddress ? `?a=${ holderAddress }` : '' }`
      : ''
  }
```

File: tsconfig.json
Comment (Gudahtt) [inline tsconfig.json:8]: Why was this changed? :thinking:

Comment (BboyAkers) [reply tsconfig.json:8]: i changed that due to the JS static method .endsWith() not being valid with es5 as the target.

**Review by Gudahtt — APPROVED** [review]
LGTM!

**Comment by BboyAkers** [comment]
@brad-decker can you approve as well por favor so I can merge? 🙂

**Review by brad-decker — APPROVED** [review]
LGTM!

**Comment by Gudahtt** [comment]
This has now been released in v2.0.0 of this package

**Pull request merged**
